extension/storage/filestorage/client.go

```diff
@@ -109,53 +128,172 @@ func (c *fileStorageClient) Batch(_ context.Context, ops ...storage.Operation) e
+		zap.String(tempDirectoryKey, file.Name()))
+
 	// cannot reuse newClient as db shouldn't contain any bucket
-	db, err := bbolt.Open(file.Name(), 0600, options)
+	compactedDb, err = bbolt.Open(file.Name(), 0600, options)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	if err := bbolt.Compact(db, c.db, maxTransactionSize); err != nil {
-		return nil, err
+	compactionStart := time.Now()
+
+	if err = bbolt.Compact(compactedDb, c.db, maxTransactionSize); err != nil {
+		return err
 	}
 
 	dbPath := c.db.Path()
-	tempDBPath := db.Path()
+	compactedDbPath := compactedDb.Path()
 
-	db.Close()
-	c.Close(ctx)
+	c.db.Close()
+	compactedDb.Close()
 
 	// replace current db file with compacted db file
-	if err := os.Remove(dbPath); err != nil {
-		return nil, err
+	if err = os.Remove(dbPath); err != nil {
+		return err
 	}
 
-	if err := os.Rename(tempDBPath, dbPath); err != nil {
-		return nil, err
+	if err = os.Rename(compactedDbPath, dbPath); err != nil {
+		return err
+	}
+
+	c.db, err = bbolt.Open(dbPath, 0600, options)
+
+	c.logger.Info("finished compaction",
+		zap.String(directoryKey, dbPath),
+		zap.Duration(elapsedKey, time.Since(compactionStart)))
+
+	return err
+}
+
+// startCompactionLoop provides asynchronous compaction function
+func (c *fileStorageClient) startCompactionLoop(ctx context.Context) {
+	ctx, c.cancel = context.WithCancel(ctx)
+
+	go func() {
+		c.logger.Debug("starting compaction loop",
+			zap.Duration("compaction_check_interval", c.compactionCfg.CheckInterval))
+
+		compactionTicker := time.NewTicker(c.compactionCfg.CheckInterval)
+		defer compactionTicker.Stop()
+
+		for {
+			select {
+			case <-compactionTicker.C:
+				if c.shouldCompact() {
+					err := c.Compact(c.compactionCfg.Directory, c.openTimeout, c.compactionCfg.MaxTransactionSize)
+					if err != nil {
+						c.logger.Error("compaction failure",
+							zap.String(directoryKey, c.compactionCfg.Directory),
+							zap.Error(err))
+					}
+				}
+			case <-ctx.Done():
+				c.logger.Debug("shutting down compaction loop")
+				return
+			}
+		}
+	}()
+}
+
+// shouldCompact checks whether the conditions for online compaction are met
+func (c *fileStorageClient) shouldCompact() bool {
+	if !c.compactionCfg.OnRebound {
+		return false
+	}
+
+	totalSize, dataSize, err := c.getDbSize()
+	if err != nil {
+		c.logger.Error("failed to get db size", zap.Error(err))
+		return false
+	}
+
+	c.logger.Debug("shouldCompact check",
+		zap.Int64("totalSize", totalSize),
+		zap.Int64("dataSize", dataSize))
+
+	if dataSize > c.compactionCfg.ReboundNeededThresholdMiB*1048576 ||
+		totalSize < c.compactionCfg.ReboundTriggerThresholdMiB*1048576 {
+		return false
+	}
+
+	c.logger.Debug("shouldCompact returns true",
+		zap.Int64("totalSize", totalSize),
+		zap.Int64("dataSize", dataSize))
+
+	return true
+}
+
+func (c *fileStorageClient) getDbSize() (totalSizeResult int64, dataSizeResult int64, errResult error) {
+	c.compactionMutex.Lock()
+	defer c.compactionMutex.Unlock()
```

Comment:
I'm not sure we need the lock here. Both readonly transactions and `.Stats()` are thread-safe in bbolt afaik. Right now we only call this in the loop, which is serial by definition.